File: algorithms/greedy/activity-selection/go/activity_selection.go
package activityselection

import "sort"

// ActivitySelection selects the maximum number of non-overlapping activities.
// The input array encodes activities as consecutive pairs [start, finish, ...].
func ActivitySelection(arr []int) int {
	n := len(arr) / 2
	if n == 0 {
		return 0
	}

	type activity struct {
		start, finish int
	}

	activities := make([]activity, n)
	for i := 0; i < n; i++ {
		activities[i] = activity{arr[2*i], arr[2*i+1]}
	}

	sort.Slice(activities, func(i, j int) bool {
		return activities[i].finish < activities[j].finish
	})

	count := 1
	lastFinish := activities[0].finish

	for i := 1; i < n; i++ {
		if activities[i].start >= lastFinish {
			count++
			lastFinish = activities[i].finish
		}
	}

	return count
}

File: algorithms/greedy/activity-selection/swift/ActivitySelection.swift
func activitySelection(_ arr: [Int]) -> Int {
    let n = arr.count / 2
    if n == 0 {
        return 0
    }

    var activities: [(start: Int, finish: Int)] = []
    for i in 0..<n {
        activities.append((start: arr[2 * i], finish: arr[2 * i + 1]))
    }

    activities.sort { $0.finish < $1.finish }

    var count = 1
    var lastFinish = activities[0].finish

    for i in 1..<n {
        if activities[i].start >= lastFinish {
            count += 1
            lastFinish = activities[i].finish
        }
    }

    return count
}
